what_about_threads/src/main.rs

```diff
@@ -0,0 +1,146 @@
+use env_logger; // 0.9.3
+use log::{info, LevelFilter, error}; // 0.4.17
+use rayon;
+use std::collections::{HashMap, HashSet};
+// 1.6.0
+use std::io::Write;
+use crossbeam_channel;
+
+// Unique identifier of work. If there are no fields work is unique.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+enum WorkIdentifier {
+    GlobalMetadata,
+    GlyphIr(u32),
+    FinishIr,
+}
+
+trait Work {
+    fn exec(&self);
+}
+
+struct StringWork {
+    message: String,
+}
+
+impl Work for StringWork {
+    fn exec(&self) {
+        info!("{}", self.message);
+    }
+}
+
+impl StringWork {
+    fn new(message: String) -> impl Work {
+        StringWork { message }
+    }
+}
+
+fn create_work() -> (HashMap<WorkIdentifier, Box<dyn Work + Send>>, HashMap<WorkIdentifier, HashSet<WorkIdentifier>>) {    
+    let mut work: HashMap<WorkIdentifier, Box<dyn Work + Send>> = HashMap::new();
+    // entry i requires entries at <indices> to complete before running
+    let mut deps: HashMap<WorkIdentifier, HashSet<WorkIdentifier>> = HashMap::new();    
+
+    // Global metadata depends on nothing and nobody!
+    work.insert(WorkIdentifier::GlobalMetadata, Box::new(StringWork::new("Global Metadata".to_string())));
+    deps.insert(WorkIdentifier::GlobalMetadata, Default::default());
+
+    for i in 0..10 {
+        let work_id = WorkIdentifier::GlyphIr(i);
+        work.insert(work_id, Box::new(StringWork::new(format!("Hi {}", i))));
+
+        // Glyphs require global metadata is done
+        deps.entry(work_id).or_default().insert(WorkIdentifier::GlobalMetadata);
+    }
+
+    // Finish IR requires all tasks except finish IR be done so take all the keys prior to adding it
+    deps.insert(WorkIdentifier::FinishIr, work.keys().into_iter().copied().collect());
+    work.insert(WorkIdentifier::FinishIr, Box::new(StringWork::new("Finish Ir".to_string())));
+
+    // It would be Very Bad if we modified this code such that this was not true
+    assert_eq!(work.keys().into_iter().collect::<HashSet<&WorkIdentifier>>(), deps.keys().into_iter().collect::<HashSet<&WorkIdentifier>>());
+    (work, deps)
+}
+
+fn main() {
+    env_logger::builder()
+        .filter(None, LevelFilter::Info)
+        .format(|buf, record| {
+            let ts = buf.timestamp_micros();
+            writeln!(
+                buf,
+                "{}: {:?}: {}: {}",
+                ts,
+                std::thread::current().id(),
+                buf.default_level_style(record.level())
+                    .value(record.level()),
+                record.args()
+            )
+        })
+        .init();
+
+    info!("main");
+    
+    let (send, recv) = crossbeam_channel::unbounded::<WorkIdentifier>();
+    rayon::in_place_scope(|scope| {
+        info!("in_place");
+        let (mut work, mut deps) = create_work();
+        let work_count = work.len();
+        let mut reverse_deps: HashMap<WorkIdentifier, Vec<WorkIdentifier>> = HashMap::new();
```

Comment:
Is that considered better? - I believe something like that did work but I didn't like it. Leave out the type, yes. Full type, yes. Partial type ... maybe I'm just not used to that option yet but it bugged me.

Reply:
my take is generally that I will provide as much information as the compiler needs, and I find really explicit type declarations feel noisy. In any case I was just pointing this out, in case it was. something you weren't aware of. :)

Reply:
`<_>` is growing on me as I use Rust more. Felt weird initially.